analyzers/tests/SonarAnalyzer.UnitTest/TestFramework/Verifier.cs

```diff
@@ -0,0 +1,106 @@
+﻿/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2021 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Reflection;
+using System.Text;
+using System.Text.RegularExpressions;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
+using SonarAnalyzer.Common;
+using SonarAnalyzer.UnitTest.Helpers;
+
+namespace SonarAnalyzer.UnitTest.TestFramework
+{
+    internal class Verifier
+    {
+        private static readonly Regex ImportsRegexVB = new(@"^\s*Imports\s+.+$", RegexOptions.Multiline | RegexOptions.RightToLeft);
+        private readonly VerifierBuilder builder;
+        private readonly DiagnosticAnalyzer[] analyzers;
+        private readonly AnalyzerLanguage language;
+
+        public Verifier(VerifierBuilder builder)
+        {
+            this.builder = builder ?? throw new ArgumentNullException(nameof(builder));
+            analyzers = builder.Analyzers.Select(x => x()).ToArray();
+            if (!analyzers.Any())
+            {
+                throw new ArgumentException(nameof(builder.Analyzers) + " cannot be empty.");
+            }
+            if (analyzers.Any(x => x == null))
+            {
+                throw new ArgumentException(nameof(builder.Analyzers) + " cannot produce null.");
+            }
+            var allLanguages = analyzers.SelectMany(x => x.GetType().GetCustomAttributes<DiagnosticAnalyzerAttribute>()).SelectMany(x => x.Languages).Distinct().ToArray();
+            if (allLanguages.Length > 1)
```

Comment:
The earlier we fail the better, can we move some of these checks to the builder?

Reply:
No, because the builder doesn't care if its intermediate state makes sense for `Verifier`. It's `Verifier`'s responsibility to do sanity checks for verifying purposes.